### PR TITLE
SPA installer should handle array-type config parameters correctly

### PR DESCRIPTION
**Pull request by brandones**
This adds support for SPA parameters that take array arguments. Right now this is just `configUrls`.

### CODE REVIEW

File: integration-tests/src/test/java/org/openmrs/maven/plugins/SetupIntegrationTest.java [hunk 1]
```diff
@@ -164,6 +164,7 @@ public void setup_shouldInstallServerFromDistroPropertiesDir() throws Exception{
         assertFilePresent(serverId + File.separator + "frontend" + File.separator + "openmrs-esm-patient-chart-app-3.0.0");
         String indexContents = FileUtils.readFileToString(new File(testDirectory.getAbsolutePath(), indexFilePath));
         assertThat(indexContents, containsString("apiUrl: \"notopenmrs\""));
+        assertThat(indexContents, containsString("configUrls: [\"foo\"]"));
```
Comment (mseaton) [inline integration-tests/src/test/java/org/openmrs/maven/plugins/SetupIntegrationTest.java:167]: It seems odd to me that we wouldn't test this with more than one array parameter, in order to test things like parsing the comma-separated syntax, etc.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java [hunk 2]
```diff
@@ -83,14 +83,33 @@ private JsonObject convertPropertiesToJSON(Map<String, String> properties) throw
         return result;
     }
 
+    /**
+     * Add a line from the properties file to the new JSON object. Creates nested objects as needed.
+     * If the value contains commas, it is assumed to represent an array of strings. Known array-valued
+     * parameters are always parsed as arrays of strings. e.g.
+     *   `foo=bar,baz` becomes `{ "foo": ["bar", "baz"] }`
+     *   `configUrl=qux` becomes `{ "configUrl": ["qux"] }` because `configUrl` is known to be array-valued
```
Comment (mseaton) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java:91]: configUrl -> configUrls

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java [hunk 3]
```diff
@@ -83,14 +83,33 @@ private JsonObject convertPropertiesToJSON(Map<String, String> properties) throw
...
+     */
     private void addPropertyToJSONObject(JsonObject jsonObject, String propertyKey, String value) {
         String[] keys = propertyKey.split("\\.");
         if (keys.length == 1) {
             if (jsonObject.has(keys[0])) {
                 throw new RuntimeException(BAD_SPA_PROPERTIES_MESSAGE +
                         " Encountered this error processing a property containing the key '" + keys[0] + "' and with value " + value);
             }
-            jsonObject.addProperty(keys[0], value);
+            if (value.contains(",") || "configUrls".equals(keys[0])) {
```
Comment (mseaton) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java:104]: Are you sure we want to make anything an array if it contains a comma?  Are there use cases where we want this, or possible use cases where we don't?

**Review by mseaton — COMMENTED** [review]
A few thoughts / suggestions.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java (hunk 3 above)
Comment (brandones) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java:104]: I have no idea, it's just speculative future-proofing. It seems more likely to me that we might add other array-valued parameters than that we will ever add a parameter which can include a `,`. But I could be wrong about that, and am open to the idea of removing this functionality.

Comment (mseaton) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java:104]: Yes, that is true, but in that case they array will also need to support single-value arrays like "configUrls", and so will also need to be hard-coded as you had to do with "configUrls", so having the generic behavior doesn't really add value or make sense.

Comment (brandones) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/utility/SpaInstaller.java:104]: Good point. Removing it.

File: integration-tests/src/test/java/org/openmrs/maven/plugins/SetupIntegrationTest.java (hunk 1 above)
Comment (brandones) [reply integration-tests/src/test/java/org/openmrs/maven/plugins/SetupIntegrationTest.java:167]: Yeah at present there's only one distro.properties file in the tests that includes the `spa.` properties, and the single array value seems more like an edge case which someone might overlook in implementing an array-values parser. Based on my experience having initially implemented this in a way that didn't support single array values :)

Could certainly add SPA building to another distro file but it would slow the tests down substantially.

Comment (ibacher) [reply integration-tests/src/test/java/org/openmrs/maven/plugins/SetupIntegrationTest.java:167]: Maybe we could test the functionality of this in a unit test rather than adding a new integration test?

Comment (brandones) [reply integration-tests/src/test/java/org/openmrs/maven/plugins/SetupIntegrationTest.java:167]: That is extremely sensible

Comment (brandones) [reply integration-tests/src/test/java/org/openmrs/maven/plugins/SetupIntegrationTest.java:167]: Added some unit tests for `SpaInstaller`. Got halfway to adding unit test support for `Setup` too, until I realized that's not what we want here. But maybe that would be useful at some point.

**Review by mseaton — APPROVED** [review]

**Review by ibacher — APPROVED** [review]
Thanks @brandones! LGTM.
